src/mame/misc/anes.cpp

```diff
@@ -38,149 +39,597 @@ namespace {
+}
+
+void anes_state::mux_w(uint8_t data)
+{
+	m_mux = data;
+	if (m_mux & (~0x1f))
+		logerror("%s: Unknown mux bit written %02X\n", machine().describe_context(), data);
+}
+
+uint8_t anes_state::key_r(offs_t offset)
+{
+	for (int i = 0; i < 5; i++)
+		if (!BIT(m_mux, i))
+			return m_key[offset][i]->read();
+
+	//  logerror("%s: Unknown key read with mux = %02X\n", machine().describe_context(), m_mux);
+	//  return 0xff;
+	return offset ? m_coin->read() : 0xff;
+}
+
+uint8_t anes_state::m1_rom_r(offs_t offset)
+{
+	uint8_t ret = m_maincpu->space(AS_PROGRAM).read_byte(offset);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_bank_delay)
+		{
+			m_banktimer->adjust(attotime::from_usec(1));
+		}
+	}
+	return ret;
+}
+
+void anes_state::palette_enable_w(uint8_t data)
+{
+	m_palette_enable = data;
+}
+
+void anes_state::set_color(int offset)
+{
+	uint16_t dat = (m_palette_data_msb[offset] << 8) | palette_data_lsb[offset];
+	m_palette->set_pen_color(offset, rgb_t(pal5bit(dat >> 0), pal5bit(dat >> 5), pal5bit(dat >> 10)));
+}
+
+void anes_state::palette_offset_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_offset_w when not enabled %02x (enable %02x)\n", data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset = data;
+	}
+}
+
+void anes_state::palette_offset_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("%s: write to palette_offset_msb_w when not enabled %02x (enable %02x)\n", machine().describe_context(), data, m_palette_enable);
+	}
+	else
+	{
+		m_palette_offset_msb = data;
+	}
+}
+
+void anes_state::palette_active_bank_w(uint8_t data)
+{
+	m_palette_active_bank = data & 0xf;
+}
+
+uint8_t anes_state::palette_data2_r()
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("read from palette_data2_r when not enabled\n");
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		return palette_data_lsb[offs];
+	}
+	return 0x00;
+}
+
+
+void anes_state::palette_data2_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data2_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		palette_data_lsb[offs] = data;
+		set_color(offs);
+	}
+}
+
+void anes_state::palette_data_msb_w(uint8_t data)
+{
+	if (m_palette_enable != 0x01)
+	{
+		logerror("write to palette_data_msb_w when not enabled %02x\n", data);
+	}
+	else
+	{
+		int offs = (m_palette_offset + (m_palette_offset_msb << 8)) & 0x1fff;
+		m_palette_data_msb[offs] = data;
+		set_color(offs);
+	}
+}
+
+
+void anes_state::opcodes_map(address_map &map)
+{
+	map(0x0000, 0xefff).r(FUNC(anes_state::m1_rom_r));
 }
 
 void anes_state::prg_map(address_map &map)
 {
-	map(0x0000, 0xefff).rom();
-	map(0xf000, 0xffff).ram();
+	map(0x0000, 0x1fff).rom();
+	map(0x2000, 0xefff).bankr(m_rombank);
+	map(0x0000, 0x01ff).w(FUNC(anes_state::blit_rom_w));
+	map(0xf000, 0xf7ff).ram();
+	map(0xf900, 0xffff).ram();
```

Comment:
Is there a real reason about this RAM hole?

Reply:
this is how it was sent to me, I think these are the ranges it tests